src/lib/protocols/beckhoff_ads.c

```diff
@@ -0,0 +1,114 @@
+/*
+ * beckhoff_ads.c
+ *
+ * Beckhoff Automation Device Specification
+ * 
+ * Copyright (C) 2023 - ntop.org
+ * Copyright (C) 2023 - V.G <jacendi@protonmail.com>
+ *
+ * This file is part of nDPI, an open source deep packet inspection
+ * library based on the OpenDPI and PACE technology by ipoque GmbH
+ *
+ * nDPI is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * nDPI is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with nDPI.  If not, see <http://www.gnu.org/licenses/>.
+ * 
+ */
+
+#include "ndpi_protocol_ids.h"
+
+#define NDPI_CURRENT_PROTO NDPI_PROTOCOL_BECKHOFF_ADS
+
+#include "ndpi_api.h"
+#include "ndpi_private.h"
+
+PACK_ON
+struct ams_tcp_hdr {
+  u_int16_t reserved;
+  u_int32_t length;
+} PACK_OFF;
+
+struct ams_hdr {
+  u_int64_t target_netid : 48;
+  u_int64_t target_port : 16;
+  u_int64_t source_netid : 48;
+  u_int64_t source_port : 16;
```

Comment:
@0xA50C1A1, the issue might be here... bitfields are endianess dependent... take a look at `src/lib/protocols/coap.c`, for an example